App Library: Vehicle Data Manager (#1061)
smartdevicelink/sdl_evolution: proposals/nnnn-app-lib-vehicle-data-manager.md

@property (nullable, strong, nonatomic) SDLDeviceStatus *deviceStatus;
@property (nullable, strong, nonatomic) SDLVehicleDataEventStatus driverBraking;
@property (nullable, strong, nonatomic) SDLWiperStatus wiperStatus;
@property (nullable, strong, nonatomic) SDLHeadLampStatus *headLampStatus;
@property (nullable, strong, nonatomic) NSNumber<SDLFloat> *engineOilLife;
@property (nullable, strong, nonatomic) NSNumber<SDLFloat> *engineTorque;
@property (nullable, strong, nonatomic) NSNumber<SDLFloat> *accPedalPosition;
@property (nullable, strong, nonatomic) NSNumber<SDLFloat> *steeringWheelAngle;
@property (nullable, strong, nonatomic) SDLECallInfo *eCallInfo;
@property (nullable, strong, nonatomic) SDLAirbagStatus *airbagStatus;
@property (nullable, strong, nonatomic) SDLEmergencyEvent *emergencyEvent;
@property (nullable, strong, nonatomic) SDLClusterModeStatus *clusterModeStatus;
@property (nullable, strong, nonatomic) SDLMyKey *myKey;
@property (nullable, strong, nonatomic) SDLElectronicParkBrakeStatus electronicParkBrakeStatus;
@property (nullable, strong, nonatomic) SDLTurnSignal turnSignal;
@property (nullable, strong, nonatomic) NSString *cloudAppVehicleID;
```

> Note that the above is a 1-1 representation of `OnVehicleData`, however, this may not always be the case. For example, with the change recommended in [SDL-0256 Refactor Fuel Information Related Data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0256-Refactor-Fuel-Information-Related-Vehicle-Data.md), the maintainers may decide that it is best to deprecate all the old elements and to coerce the old data types into the new data structure automatically.

Reviewer: ```suggestion
Note that the above is a 1-1 representation of `OnVehicleData`, however, this may not always be the case. For example, with the change recommended in [SDL-0256 Refactor Fuel Information Related Data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0256-Refactor-Fuel-Information-Related-Vehicle-Data.md), the SDLC Project Maintainer may decide that it is best to deprecate all the old elements and to coerce the old data types into the new data structure automatically.
```

When data is subscribed, the developer will receive updates on a regular basis as the vehicle data value updates. It uses the `SubscribeVehicleData` and `UnsubscribeVehicleData` RPCs under the hood.

#### iOS
```objc
/// A handler to be called when vehicle data unsubscribes
///
/// @param error An error if one occurred while unsubscribing
typedef void (^SDLVehicleDataUnsubscribeHandler)(NSError *_nullable error);

/// Subscribes to one or more pieces of vehicle data and calls a handler when any of the types is updated. It will call the handler individually for each update. Vehicle data will be subscribed individually under-the-hood, meaning that the if an error is called, it only relates to that one vehicle data type, but that the callback may be called multiple times with multiple errors.
///
/// @param types The vehicle data types to subscribe to for this handler
/// @param handler The handler to call every time any of the specified vehicle data types update
/// @return A unique object that can be passed to `unsubscribeFromVehicleData:withObserver:`
- (id<NSObject>)subscribeToVehicleData:(NSArray<SDLVehicleDataType>)types withUpdateHandler:(SDLVehicleDataUpdateHandler)handler;

> /// Subscribes to one or more pieces of vehicle data and calls a selector on an observer when any of these types is updated. It will call the handler individually for each update. Vehicle data will be subscribed individually under-the-hood, meaning that the if an error is called, it only relates to that one vehicle data type, but that the callback may be called multiple times with multiple errors.

Reviewer: ```suggestion
/// Subscribes to one or more pieces of vehicle data and calls a selector on an observer when any of these types is updated. It will call the handler individually for each update. Vehicle data will be subscribed individually under the hood, meaning that if an error is called, it only relates to that one vehicle data type, but that the callback may be called multiple times with multiple errors.
```

///

> /// The selector must exist with between 0 and 3 parameters. If zero parameters, the selector will be called, but you will be unable to determine which type was updated or know about any errors that occur. If one parameter, the parameter must be of type `SDLVehicleDataType`, which is the type that was updated. If two parameters, the second parameter must of type `NSError` and will be an error that occurs when subscribing to the `types`. If three parameters, the third parameter must be of type `SDLVehicleDataManager`. For example, `vehicleDataTypeDidUpdate:(SDLVehicleDataType)type error:(NSError *)error manager:(SDLVehicleDataManager *)manager`.

Reviewer: ```suggestion
/// The selector must exist with between 0 and 3 parameters. If zero parameters, the selector will be called, but you will be unable to determine which type was updated or know about any errors that occur. If one parameter, the parameter must be of type `SDLVehicleDataType`, which is the type that was updated. If two parameters, the second parameter must be of type `NSError` and will be an error that occurs when subscribing to the `types`. If three parameters, the third parameter must be of type `SDLVehicleDataManager`. For example, `vehicleDataTypeDidUpdate:(SDLVehicleDataType)type error:(NSError *)error manager:(SDLVehicleDataManager *)manager`.
```
